### DISCUSSION TITLE
Spack install Sw4 ubuntu 22.4.3

### DISCUSSION THREAD

**Discussion by ghost** [geodynamics/sw4#196] [Q&A]
hi,

while install on ubuntu 22.4.3 i got breaks

" /usr/bin/ld: main.o: in function main': main.C:(.text.startup+0x239): undefined reference to H5Z_zfp_initialize’ /usr/bin/ld: main.C:(.text.startup+0x324): undefined reference to `H5Z_zfp_final ize’

cheers,

**Comment by houjun** [comment]
Hi @ronws, looks like there is something wrong with the H5Z-ZFP library installation that SW4 depends on, can you verify that H5Z-ZFP is installed? e.g. "spack list h5z-zfp"
Or if you do not need to use the ssioutput with compression, try installing SW4 with "spack install sw4~zfp"

**Comment by ghost** [comment]
>> 332    main.C:(.text.startup+0x239): undefined reference to `H5Z_zfp_initialize'
  >> 333    /usr/bin/ld: main.C:(.text.startup+0x324): undefined reference to `H5Z_zfp_finalize'
     334    /usr/bin/ld: ESSI3DHDF5.o: in function `ESSI3DHDF5::init_write_vel(bool, int, int, double, int)':
  >> 335    ESSI3DHDF5.C:(.text+0xfbc): undefined reference to `H5Pset_zfp_reversible'
  >> 336    /usr/bin/ld: ESSI3DHDF5.C:(.text+0xfd4): undefined reference to `H5Pset_zfp_accuracy'
  >> 337    /usr/bin/ld: ESSI3DHDF5.C:(.text+0xfed): undefined reference to `H5Pset_zfp_precision'
  >> 338    /usr/bin/ld: ESSI3DHDF5.C:(.text+0x100c): undefined reference to `H5Pset_zfp_rate'
     339    /usr/bin/ld: CheckPoint.o: in function `CheckPoint::write_checkpoint_hdf5(double, int, std::vector<Sarray, std::allocator<Sarray> >&, std::vector<Sarray, std::allocator<Sarray> >&, std::vector<Sarra
            y*, std::allocator<Sarray*> >&, std::vector<Sarray*, std::allocator<Sarray*> >&)':
  >> 340    CheckPoint.C:(.text+0xa061): undefined reference to `H5Pset_zfp_reversible'
  >> 341    /usr/bin/ld: CheckPoint.C:(.text+0xa07b): undefined reference to `H5Pset_zfp_accuracy'
  >> 342    /usr/bin/ld: CheckPoint.C:(.text+0xa095): undefined reference to `H5Pset_zfp_precision'
  >> 343    /usr/bin/ld: CheckPoint.C:(.text+0xa0af): undefined reference to `H5Pset_zfp_rate'
  >> 344    collect2: error: ld returned 1 exit status
  >> 345    make: *** [Makefile:304: sw4] Error 1

**Comment by ghost** [comment]
>> 26    ar: libh5zzfp.a: error reading H5Zzfp_props.o: file truncated
  >> 27    make[1]: *** [Makefile:44: libh5zzfp.a(H5Zzfp_props_f.o)] Error 1
     28    make[1]: *** Waiting for unfinished jobs....
     29    make[1]: Leaving directory '/tmp/tech/spack-stage/spack-stage-h5z-zfp-develop-7uhkj4qsuxmnpnwmo2jeicmnon7gbpf5/spack-src/src'
  >> 30    make: *** [Makefile:44: all] Error 2

the h5z-zfp installed

**Reply by houjun** [reply]
Looks like your H5Z-ZFP library installation failed, can you try to uninstall it and then install again? 
spack uninstall h5z-zfp
spack install h5z-zfp
Alternatively, you can create a new spack environment and redo the sw4 installation from scratch.

**Reply by ghost** [reply]
898        # dependencies depends_on('mpi'), require Fortran compiler to
        899        # avoid delayed build errors in dependents.
        900        if (self.compiler.f77 is None) and (self.compiler.fc is None):
  >>    901            raise InstallError("OpenMPI requires both C and Fortran compilers!")

**Reply by houjun** [reply]
It looks like your system doesn't have a valid Fortran compiler, causing the error. 
Try "which gfortran" on your system, and if there is none available, use one of the following command to install it:
sudo apt install gfortran (Debian-based)
or
sudo yum install gcc-gfortran (RPM-based)

**Comment by ghost** [comment]
>> 26    ar: libh5zzfp.a: error reading H5Zzfp_props.o: file truncated
  >> 27    make[1]: *** [Makefile:44: libh5zzfp.a(H5Zzfp_props_f.o)] Error 1
     28    make[1]: *** Waiting for unfinished jobs....
     29    make[1]: Leaving directory '/tmp/dev/spack-stage/spack-stage-h5z-zfp
           -develop-p2y2z4tqsehfexwn7425ufh7s5bnudyg/spack-src/src'
  >> 30    make: *** [Makefile:44: all] Error 2

**Reply by houjun** [reply]
Something went wrong with the H5Z-ZFP installation, can you make a new spack environment and just do "spack install H5Z-ZFP" and see if the install is successful. If not, I'd suggest contacting the H5Z-ZFP developers for their help.
Have you tried installing sw4 without zfp?

**Comment by ghost** [comment]
hi,

same same 

  >> 26    ar: libh5zzfp.a: error reading H5Zzfp_props.o: file truncated
  >> 27    make[1]: *** [Makefile:44: libh5zzfp.a(H5Zzfp_props_f.o)] Error 1
     28    make[1]: *** Waiting for unfinished jobs....
     29    make[1]: Leaving directory '/tmp/dev/spack-stage/spack-stage-h5z-zfp
           -1.1.1-sefdj72m2eocdkzzmfdg73u4dytvs3f6/spack-src/src'
  >> 30    make: *** [Makefile:44: all] Error 2


i've notice even though the source control is version 3
when install within spack its still versions beta 3

**Reply by houjun** [reply]
I just created a pull request to have spack install SW4 3.0, it should be merged in the next couple of days.
For now you can try with "spack install sw4@master" to install the latest SW4 code, but I think you will encounter the same issue as it is H5Z-ZFP related. Please contact H5Z-ZFP developers to see if they can resolve the problem.

**Comment by ghost** [comment]
same same.
  >> 26    ar: libh5zzfp.a: error reading H5Zzfp_lib.o: file truncated
  >> 27    make[1]: *** [Makefile:40: libh5zzfp.a(H5Zzfp_props.o)] Error 1
     28    make[1]: *** Waiting for unfinished jobs....
     29    make[1]: Leaving directory '/tmp/tech/spack-stage/spack-stage-h5z-zf
           p-develop-7uhkj4qsuxmnpnwmo2jeicmnon7gbpf5/spack-src/src'
  >> 30    make: *** [Makefile:44: all] Error 2

See build log for details:
  /tmp/tech/spack-stage/spack-stage-h5z-zfp-develop-7uhkj4qsuxmnpnwmo2jeicmnon7gbpf5/spack-build-out.txt

==> Warning: Skipping build of sw4-master-c7vlqrx54sd773ghvup2zangcn4h26uc since h5z-zfp-develop-7uhkj4qsuxmnpnwmo2jeicmnon7gbpf5 failed
[+] /home/tech/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-h5py-3.8.0-bedfnrd7fbxal2o63rvjwsa7tzc3hhnu
==> Error: sw4-master-c7vlqrx54sd773ghvup2zangcn4h26uc: Package was not installed
==> Error: Installation request failed.  Refer to reported errors for failing p
thank you.

**Comment by ghost** [comment]
Hi all,

ive installed as suggested. 
h5z-zfp

yet seems same same . 
Error: KeyError: 'No spec with name h5z-zfp in sw4@=3.0%gcc@=11.4.0+fftw+hdf

>     51        os.environ["H5ZROOT"] = spec["h5z-zfp"].prefix

**Reply by houjun** [reply]
It looks like Spack installation is not working for you, since it depends a lot on your system environment, it is difficult for me to figure out what went wrong. So I'd suggest installing SW4 with the Makefile approach as described in the Installation Guide, you can still try using the dependent libraries (HDF5, proj, etc.) installed by Spack by specifying their paths in the configs/make.inc file, but compile SW4 from the source code.
